File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Solution
{
    public int solution(int[,] maps)
    {
        int Row = maps.GetLength(0);
        int Column = maps.GetLength(1);

        int[] DirX = { -1, 1, 0, 0 };
        int[] DirY = { 0, 0, -1, 1 };

        int[,] Pay = new int[Row, Column];
        bool[,] IsVisited = new bool[Row, Column];

        Queue<(int, int)> BFS = new Queue<(int, int)>();

        BFS.Enqueue((0, 0));
        Pay[0, 0] = 1;
        IsVisited[0, 0] = true;

        while (BFS.Count > 0)
        {
            var (curX, curY) = BFS.Dequeue();

            if (curX == Row - 1 && curY == Column - 1)
            {
                return Pay[curX, curY];
            }

            for (int i = 0; i < 4; ++i)
            {
                int NewX = curX + DirX[i];
                int NewY = curY + DirY[i];

                if (NewX < 0 || NewY < 0 || NewX == Row || NewY == Column 
                    || IsVisited[NewX, NewY] || maps[NewX, NewY] == 0)
                {
                    continue;
                }

                IsVisited[NewX, NewY] = true;
                Pay[NewX, NewY] = Pay[curX, curY] + 1;
                BFS.Enqueue((NewX, NewY));
            }
        }

        return -1;
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include <vector>
#include <queue>
#include <array>
#include <iostream>

int solution(std::vector<std::vector<int>> maps)
{
    int answer = 0;
    
    const size_t Row = maps.size();
    const size_t Column = maps[0].size();
    
    std::array<int, 4> DirX = {-1, 1, 0, 0};
    std::array<int, 4> DirY = {0, 0, -1, 1};
    
    std::vector<std::vector<int>> Pay(Row, std::vector<int>(Column, 0));
    std::vector<std::vector<bool>> IsVisited(Row, std::vector<bool>(Column, false));
    std::queue<std::pair<int, int>> BFS;
    
    BFS.push(std::make_pair(0, 0));
    IsVisited[0][0] = true;
    Pay[0][0] = 1;
    
    while(false == BFS.empty())
    {
        std::pair<int, int> CurNode = BFS.front();
        BFS.pop();
        
        for(int i = 0; i < 4; ++i)
        {
            int NewX = CurNode.first + DirX[i];
            int NewY = CurNode.second + DirY[i];
            
            if(NewX < 0 || NewY < 0 || NewX >= Row || NewY >= Column)
            {
                continue;
            }
            
            if(0 == maps[NewX][NewY])
            {
                continue;
            }
            
            if(true == IsVisited[NewX][NewY])
            {
                continue;
            }
            
            IsVisited[NewX][NewY] = true;
            Pay[NewX][NewY] = Pay[CurNode.first][CurNode.second] + 1;
            BFS.push(std::make_pair(NewX, NewY));
        }
    }
    
    if(true == IsVisited[Row - 1][Column - 1])
    {
        answer = Pay[Row - 1][Column - 1];
    }
    else
    {
        answer = -1;
    }
    
    return answer;
}